runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableUpcallHandler.cpp

```diff
@@ -0,0 +1,151 @@
+	if (NULL == upcallMetaData) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+
+	nativeSig = (J9UpcallNativeSignature *)j9mem_allocate_memory(sizeof(J9UpcallNativeSignature), OMRMEM_CATEGORY_VM);
+	if (NULL == nativeSig) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	sigArray = (J9UpcallSigType *)j9mem_allocate_memory(sizeof(J9UpcallSigType) * sigCount, OMRMEM_CATEGORY_VM);
+	if (NULL == sigArray) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	for (U_32 sigIndex = 0; sigIndex < sigCount; sigIndex++) {
+		j9object_t sigStrObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, cSigStrs, sigIndex);
+		char sigBuffer[J9VM_NATIVE_SIGNATURE_STRING_LENGTH] = {0};
+		/* The simplified signature string in cSig for parameter/return type is converted at java level.
+		 * e.g.
+		 * "4#I" represents a 4-byte integer
+		 * "20#[5:I]" represents a 20-byte struct for {int a[5]}
+		 * "16#[I(4)J]" represents a 16-byte struct for {int, padding(4 bytes), long}
+		 *
+		 * Note: the last element is the signature string for the return type.
+		 */
+		char *cSig = copyStringToUTF8WithMemAlloc(currentThread, sigStrObject,
+				J9_STR_NULL_TERMINATE_RESULT, "", 0, sigBuffer, sizeof(sigBuffer), NULL);
+		LayoutFFITypeHelpers::encodeUpcallSignature(cSig, &sigArray[sigIndex]);
+		if (cSig != sigBuffer) {
+			j9mem_free_memory(cSig);
+		}
+	}
+	nativeSig->numSigs = sigCount;
+	nativeSig->sigArray  = sigArray;
+
+	/* Set the fields of the J9UpcallMetaData struct for the thunk generation */
+	upcallMetaData->vm = vm;
+	upcallMetaData->nativeFuncSignature = nativeSig;
+	mhMetaData = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 1);
+	upcallMetaData->mhMetaData = j9jni_createGlobalRef((JNIEnv*)currentThread, mhMetaData, false);
+
+	/* The resolution is performed by MethodHandleResolver.upcallLinkCallerMethod()
+	 * to obtain a MemberName object plus appendix in a two element object array.
+	 */
+	VM_OutOfLineINL_Helpers::buildInternalNativeStackFrame(currentThread, method);
+	resolveUpcallInvokeHandle(currentThread, upcallMetaData);
+	if (VM_VMHelpers::exceptionPending(currentThread)) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		goto freeAllMemoryThenExit;
+	}
+	VM_OutOfLineINL_Helpers::restoreInternalNativeStackFrame(currentThread);
+
+	// Comment out temporarily as the code involved is split up in part 3 (thunk generation)
+	// thunkAddr = vmFuncs->createUpcallThunk(upcallMetaData);
+	if (NULL == thunkAddr) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+done:
+	VM_OutOfLineINL_Helpers::returnDouble(currentThread, (intptr_t)thunkAddr, 3);
+	return rc;
+
+freeAllMemoryThenExit:
+	if (NULL != upcallMetaData) {
```

Comment:
the NULL check isnt necesarry as `j9mem_free_memory` will do that

Reply:
Already updated in `/runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler.cpp`.